README
DBQuery version 1.000
===============

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test

if some wrong here when make test, you should modify t/DBQuery.t which contains db_user or db_pass etc.

   make install

DEPENDENCIES

This module requires these other modules and libraries:

  DBI
  DBD:mysql
  DBD::PgSQL
  DBD::Oracle

COPYRIGHT AND LICENCE

for Yahoo. and Alibaba.

HELP

You can get help from the order:

pod2text lib/DBQuery.pm
perldoc lib/DBQuery.pm

Copyright (C) 2009 by cnangel

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension DBQuery.

1.10  Thu Jul 07 11:18:13 2011 
    - Add compatibility about oracle and mysql;
    - Update documents and examples.
1.00  Fri Apr 02 02:58:16 2010
    - Add function reconnect;
    - Add function autocommit;
    - Add function charset store.
0.08  Wed Mar 31 12:55:00 2010
    - Add Charset for connection.  
0.07  Thu Dec 03 11:00:35 2009
    - Add a method about reference of array;
    - Add pod document for method.
0.06  Wed Dec 02 11:41:50 2009 
    - Replace perldoc DBQuery with pod document.
0.05  Tue Jul 07 10:19:28 2009
	- Repair and pass test.
0.04  Tue Jun 30 17:31:44 2009
	- Modify the test case.
0.03  Mon Jun 29 16:53:04 2009
	- Can miss out the arg: db_host, and optimize the test
0.02  Mon Jun 29 10:22:35 2009
	- Modify name: DB to DBQuery.
0.01  Wed Jun 24 17:08:43 2009
	- Original version created by cnangel.



t/DB.pm
package t::DB;

use Test::Base -Base;
use IPC::Run3 (); 
use FindBin;

our @EXPORT = qw( run_tests );

$ENV{LC_ALL} = 'C';

sub run_tests ()
{
    for my $block (blocks())
	{
        run_test($block);
    }
}

sub run_test ($)
{
}

1;


t/DBQuery.pm
package t::DBQuery;

use Test::Base -Base;
#use IPC::Run3 ();
use FindBin;
use Data::Dumper;

our @EXPORT = qw( run_tests );

$ENV{LC_ALL} = 'C';
my $is_linux = ($^O =~ /linux/i);

sub run_tests ()
{
    for my $block (blocks())
	{
        run_test($block);
    }
}

sub run_test ($)
{
    my $block = shift;
    my $name = $block->name;

    if (defined $block->linux_only && ! $is_linux)
	{
        diag "$name - Tests skipped on $^O\n";
        for (1..3)
		{
            pass("tests skipped on $^O\n");
        }
        return;
    }

	if (defined $block->connect)
	{
		eval { use DBQuery; };
		is(0, ($@ ? 1 : 0), "$name - status ok");
		my $cmd = "mysql -u" . $block->db_user . ($block->db_pass ? " -p" . $block->db_pass : "") . " " . $block->db_name . " -e 'show tables'";
		#eval { qx($cmd); };
		my $ret = system($cmd);
		#if ($@) 
		if ($ret)
		{
			pass("tests skipped because of username or password\n");
			return;
		}
		diag "\n" . $block->db_user . ' as database user';
		diag $block->db_pass . ' as database password';
#		warn Dumper $block->original_values;
		my $db = new DBQuery($block->original_values);
		eval { $db->connect(); };
		is(0, ($@ ? 1 : 0), "$name - status ok");
	}

#	my ($in, $out, $err);
#   IPC::Run3::run3 $cmd, \$in, \$out, \$err;
#	if (defined $block->err) 
#	{
#		$err =~ s/\Q$RcFile\E/**RC_FILE_PATH**/g;
#		is $err, $block->err, "$name - stderr ok";
#	}
#	elsif ($err)
#	{
#		warn $err, "\n";
#	}   

#	if (defined $block->out) 
#	{
#		$out =~ s/\Q$RcFile\E/**RC_FILE_PATH**/g;
#		is $out, $block->out, "$name - stdout ok";
#	}   
}

1;



Makefile.PL
use strict;
use lib '.';
use inc::Module::Install;

name                ('DBQuery');
license             ('bsd');
author              ('Cnangel <junliang.li@alibaba-inc.com>');
perl_version        ('5.006001');
all_from            ('lib/DBQuery.pm');

requires            ('DBI'   => '1.4');

build_requires      ('Test::Base' => '0.10');

use_test_base;

no_index( directory => qw< t inc > );

auto_install();
WriteAll();

=head1 OLD

use 5.006001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'DBQuery',
    VERSION_FROM      => 'lib/DBQuery.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (#ABSTRACT       => 'DBQuery',
       ABSTRACT_FROM  => 'lib/DBQuery.pm', # retrieve abstract from module
       AUTHOR         => 'Cnangel <junliang.li@alibaba-inc.com>') : ()),
);

=cut


lib/DBQuery.pm
package DBQuery;

# $Id: DBQuery.pm,v 1.0.0-0 2011/07/07 11:02:57 Cnangel Exp $

use DBI;

$DBQuery::VERSION = "1.100";

sub new
{
	my $class = shift;
	my $DB = shift;
	my $self;
	if (defined $DB->{dsn})
	{
		$self = {
			'dsn' => $DB->{dsn},
			'user' => $DB->{db_user},
			'pass' => (defined $DB->{db_pass} ? $DB->{db_pass} : ''),
			'dbh' => undef,
			'sth' => undef,
		};
	}
	else
	{
		$DB->{driver_name} = 'mysql' unless (defined $DB->{driver_name});
		$DB->{driver_name} = ucfirst($DB->{driver_name}) if ($DB->{driver_name} eq "oracle");
		$self = {
			'driver' => $DB->{driver_name},
			'dsn' => $DB->{driver_name} eq 'mysql'
				? 'dbi:' . $DB->{driver_name} . ':database=' . $DB->{db_name} .
				(defined $DB->{db_host} ? ';host=' . $DB->{db_host} : '') .
				(defined $DB->{db_sock} ? ';mysql_socket=' . $DB->{db_sock} : ';mysql_socket=/var/lib/mysql/mysql.sock') . 
				(defined $DB->{db_port} ? ';port=' . $DB->{db_port} : ';port=3306') 
				: ($DB->{driver_name} eq 'pgsql' 
						? 'dbi:' . $DB->{driver_name} . ':dbname=' . $DB->{db_name} . '' . 
						(defined $DB->{db_host} ? ';host=' . $DB->{db_host} : '') . 
						(defined $DB->{db_path} ? ';path=' . $DB->{db_path} : '') .
						(defined $DB->{db_port} ? ';port=' . $DB->{db_port} : ';port=5432')
						: ($DB->{driver_name} eq 'Oracle'
							? 'dbi:' . $DB->{driver_name} . 
							(defined $DB->{db_host} ? ':host=' . $DB->{db_host} : ':host=localhost') . 
							(defined $DB->{db_port} ? ';port=' . $DB->{db_port} : '') .
							(defined $DB->{db_sid} ? ';sid=' . $DB->{db_sid} : '') .
							(defined $DB->{db_name} && !defined $DB->{db_sid} ? ';sid=' . $DB->{db_name} : '')
							: 'dbi:' . $DB->{driver_name} . (defined $DB->{db_host} ? ':' . $DB->{db_host} : '')
						  )
				  ),
			'user' => $DB->{db_user},
			'pass' => (defined $DB->{db_pass} ? $DB->{db_pass} : ''),
			'pconnect' => $DB->{db_pconnect},
			'utf8' => $DB->{db_enable_utf8},
			'autocommit' => (defined $DB->{db_autocommit} ? $DB->{db_autocommit} : 1),
			'LongReadLen' => $DB->{db_longreadlen},
			'LongTruncOk' => $DB->{db_longtruncok},
			'dbh' => undef,
			'sth' => undef,
		};
	}
	bless $self, $class;
	return $self;
}

sub connect
{
	my $self = shift;
	if ($_[0] && $self->{driver} eq 'mysql') {
		$self->{dbh} = DBI->connect($self->{dsn}, $self->{user}, $self->{pass}, {'RaiseError' => 1, 'mysql_enable_utf8' => 1});
	} else {
		$self->{dbh} = DBI->connect($self->{dsn}, $self->{user}, $self->{pass}, {'RaiseError' => 1});
	}

	if ($self->{driver} eq 'mysql') {
		$self->{dbh}->{mysql_auto_reconnect} = $self->{pconnect} ? 1 : 0;
		$self->{dbh}->{mysql_enable_utf8} = $self->{utf8} ? 1 : 0;
		$self->{dbh}->{mysql_no_autocommit_cmd} = $self->{autocommit} ? 0 : 1;
	} elsif ($self->{driver} eq 'Oracle') {
		$self->{dbh}->{LongReadLen} = $self->{LongReadLen};
		$self->{dbh}->{LongTruncOk} = $self->{LongTruncOk};
	}
	return;
}

sub query
{
	my $self = shift;
	$self->{sth} = $self->{dbh}->prepare($_[0]);
	$self->{sth}->execute();
	return $self->{sth};
}

sub quote
{
	my $self = shift;
	return $self->{dbh}->quote($_[0]);
}

sub fetch_array
{
	my $self = shift;
	return ref($_[0]) eq 'DBI::st' ? $_[0]->fetchrow_array() : $self->{sth}->fetchrow_array();
}

sub fetch_arrayref
{
	my $self = shift;
	return ref($_[0]) eq 'DBI::st' ? $_[0]->fetchrow_arrayref() : $self->{sth}->fetchrow_arrayref();
}

sub fetch_hash
{
	my $self = shift;
	return ref($_[0]) eq 'DBI::st' ? $_[0]->fetchrow_hashref() : $self->{sth}->fetchrow_hashref();
}

sub close
{
	my $self = shift;
	$self->{sth}->finish() if (defined $self->{sth});
	$self->{dbh}->disconnect if (defined $self->{dbh});
	return;
}

1;

__END__

=head1 NAME

DBQuery - Lib of DB Query

=head1 SYNOPSIS

The following lib are provided:

=over

=item B<import>

use DBQuery;

=item B<Struct Init>

Init mysql struct example:

    my %DB = (
		'db_host'			=> 'web10.search.cnb.yahoo.com',
		'db_user'			=> 'yahoo',
		'db_pass'			=> 'yahoo',
		'db_name'			=> 'ADCode',
		'db_port'			=> 3306,
		'db_pconnect'			=> 1,
		'db_autocommit'			=> 1,
		'db_enable_utf8'		=> 0,
		);
    my $db = new DBQuery(\%DB);

or postgresql:

    my %PQ = (
		'driver_name'		=> 'PgPP',
		'db_host'		=> 'tool2.search.cnb.yahoo.com',
		'db_name'		=> 'cnedb',
		'db_user'		=> 'cnedb',
		'db_pass'		=> 'cnedb',
		);
    my $db = new DBQuery(\%PQ);

or oracle:

    my %OC = (
		'driver_name'			=> 'oracle',
		'db_host'			=> 'ocndb',
		'db_user'			=> 'alibaba',
		'db_pass'			=> 'ocndb',
		'db_port'			=> 1521,
		'db_name'			=> 'ctutest', // the same as db_sid
		'db_longreadlen'		=> 33554432,
		'db_longtruncok'		=> 1,
		);
    my $db = new DBQuery(\%OC);

over this, you can use dsn for init structure.

    my %DB = (
		'dsn'		=> 'dbi:mysql:database=testinter;host=localhost;mysql_socket=/var/lib/mysql/mysql.sock;mysql_use_result=1',
		'db_user'       => 'pca',
		'db_pass'       => 'pca',
		);
    my $db = new DBQuery(\%DB);

it yet run.

=item B<Connect>

Connect resource from database.

    $db->connect();

You can unset the variable: %DB, %PQ or $OC, like this:

    undef %PQ;

or 

    undef %DB;

or

    undef %OC;


=item B<Query>

Simple query:

    $db->query("select url from edb.white_black_grey where spamtype=':demote2:' limit 10;");
    while (my @row = $db->fetch_array())
    {
    	print Dumper @row, "\n";
    }

    $db->query("alter session set nls_date_format = 'yyyy-mm-dd hh24:mi:ss'");

Common:

    my $query = $db->query("select url from edb.white_black_grey where spamtype=':demote2:' limit 10;");
    while (my @row = $db->fetch_array($query))
    {
    	print Dumper @row, "\n";
    }

=item B<Disconnect>

Release resource from database.

    $db->close();

=back

=head1 OPTIONS

Nothing because of no script.

=head1 DESCRIPTION

C<DBQuery> allows you to query some information from some different type databases, like mysql, postgresql and oracle, so our system need module which include L<DBD::mysql>, L<DBD::PgPP> and L<DBD::Oracle>.

In furture, it'll support more and more database types if you want. You can use C<DBQuery> very expediently. so we use database easily.

B<This lib> can use dsn which contains all connection infomation or use all single items, like db_host, db_pass etc.

=head2 $self->new()

Construct.

=head2  $self->connect()

Create a connect.

=head2  $self->close()

Close this connection.

=head2 $self->query()

Send a query.

=head2 $self->fetch_array()

Fetch and return array.

=head2 $self->fetch_arrayref()

Fetch and return reference of array.

=head2 $self->fetch_hash()

Fetch and return reference of hash.

=head2 $self->quote()

Quote some characters.

=head1 TIPS

There's some extra tips found in our own's everyday use:

=over

=item B<CUSTOM>

Like php-mysql

   $db->connect();
   $db->query();
   $db->fetch_array();

=item B<ABSTRACT>

One line description of the module. Will be included in PPD file.

=item B<ABSTRACT_FROM>

Name of the file that contains the package description. MakeMaker looks
for a line in the POD matching /^($package\s-\s)(.*)/. This is typically
the first line in the "=head1 NAME" section. $2 becomes the abstract.

=back

=head1 PREREQUISITES

This module uses L<DBI>.

=head1 INSTALLATION

If you are not soudoer or root, you need contact administrator.

    perl Makefile.PL
    make
    make test
    make install

Win32 users should replace "make" with "nmake".

=head1 SOURCE CONTROL

You can always get the latest SSH::Batch source from its
public Git repository:

    http://github.com/cnangel/DB/tree/master

If you have a branch for me to pull, please let me know ;)

=head1 TODO

To this:

=over

=item *

Sqlite2 and sqlite3 will be supported.

=item *

New engine for DBQuery.

=back

=head1 SEE ALSO

L<DBI>, L<DBD::mysql>, L<DBD::PgPP>, L<DBD::Oracle>

=head1 COPYRIGHT AND LICENSE

This module as well as its programs are licensed under the BSD License.

Copyright (c) 2007-2008, Yahoo! China Relevance Team, Alibaba Inc. All rights reserved.

Copyright (c) 2009, Alibaba Search Center, Alibaba Inc. All rights reserved.

Copyright (C) 2009, Cnangel Li (cnangel). All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

=over

=item *

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

=item *

Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

=item *

Neither the name of the Alibaba Search Center, Alibaba Inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

=back

=head1 AUTHOR 

B<Cnangel> (I<junliang.li#alibaba-inc.com>)

=head1 HISTORY

see ChangeLog.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



lib/DB.pm
package DB;

use DBI;

$DB::VERSION = 0.004;

sub new
{
	my $class = shift;
	my $DB = shift;
	my $self;
	if (defined $DB->{dsn})
	{
		$self = {
			'dsn' => $DB->{dsn},
			'user' => $DB->{db_user},
			'pass' => $DB->{db_pass},
			'dbh' => undef,
			'sth' => undef,
		};
	}
	else
	{
		$DB->{driver_name} = 'mysql' unless (defined $DB->{driver_name});
		$DB->{driver_name} = ucfirst($DB->{driver_name}) if ($DB->{driver_name} eq "oracle");
#	$DB->{db_port} = 3306 unless (defined $DB->{db_port});
#	$DB->{db_sock} = '/tmp/mysql.sock' unless (defined $DB->{db_sock});
		$self = {
			'dsn' => $DB->{driver_name} eq 'mysql'
				? 'dbi:' . $DB->{driver_name} . ':database=' . $DB->{db_name} . ';hostname=' . $DB->{db_host} . ';' . (defined $DB->{db_sock} ? 'mysql_socket=' . $DB->{db_sock} . ';' : 'mysql_socket=/tmp/mysql.sock;') . (defined $DB->{db_port} ? 'port=' . $DB->{db_port} : 'port=3306') 
				: ($DB->{driver_name} eq 'pgsql' 
						? 'dbi:' . $DB->{driver_name} . ':dbname=' . $DB->{db_name} . ';' . (defined $DB->{db_host} ? 'host=' . $DB->{db_host} . ';' : '') . (defined $DB->{db_path} ? 'path=' . $DB->{db_path} . ';' : '') . (defined $DB->{db_port} ? 'port=' . $DB->{db_port} : 'port=5432')
						: 'dbi:' . $DB->{driver_name} . (defined $DB->{db_host} ? ':' . $DB->{db_host} : '')
				  ),
			'user' => $DB->{db_user},
			'pass' => $DB->{db_pass},
			'dbh' => undef,
			'sth' => undef,
		};
	}
	bless $self, $class;
	return $self;
}

sub connect
{
	my $self = shift;
	$self->{dbh} = DBI->connect($self->{dsn}, $self->{user}, $self->{pass}, {'RaiseError' => 1});
	return;
}

sub query
{
	my $self = shift;
	$self->{sth} = $self->{dbh}->prepare($_[0]);
	$self->{sth}->execute();
	return $self->{sth};
}

sub quote
{
	my $self = shift;
	return $self->{dbh}->quote($_[0]);
}

sub fetch_array
{
	my $self = shift;
	return ref($_[0]) eq 'DBI::st' ? $_[0]->fetchrow_array() : $self->{sth}->fetchrow_array();
}

sub fetch_hash
{
	my $self = shift;
	return ref($_[0]) eq 'DBI::st' ? $_[0]->fetchrow_hashref() : $self->{sth}->fetchrow_hashref();
}

sub close
{
	my $self = shift;
	$self->{sth}->finish() if (defined $self->{sth});
	$self->{dbh}->disconnect if (defined $self->{dbh});
	return;
}

1;

__END__

=head1 NAME

DB.pm  - Lib of DB Query

=head1 SYNOPSIS

use DB;

=head1 OPTIONS

=over 8

=item B<Struct Init>

Mysql example:

my %DB = (
		'db_host'	=> 'web10.search.cnb.yahoo.com',
		'db_user'	=> 'yahoo',
		'db_pass'	=> 'yahoo',
		'db_name'	=> 'ADCode',
		);

my $db = new DB(\%DB);

or postgresql example:

my %PQ = (
		'driver_name'	=> 'PgPP',
		'db_host'		=> 'tool2.search.cnb.yahoo.com',
		'db_name'		=> 'cnedb',
		'db_user'		=> 'cnedb',
		'db_pass'		=> 'cnedb',
		);

or oracle example:

my %OC = (
		'driver_name'	=> 'oracle',
		'db_host'		=> 'ocndb',
		'db_user'		=> 'alibaba',
		'db_pass'		=> 'ocndb',
		);

my $db = new DB(\%OC);

over this, you can use dsn for init structure.

my %DB = (
		'dsn'			=> 'dbi:mysql:database=testinter;host=localhost;mysql_socket=/var/lib/mysql/mysql.sock',
		'db_user'       => 'pca',
		'db_pass'       => 'pca',
		);

it yet run.

=item B<Connect>


$db->connect();

You can unset the variable: %DB, %PQ or $OC, like this:

undef %PQ;
or 
undef %DB;
or
undef %OC;

=item B<Query>


Simple:

$db->query("select url from edb.white_black_grey where spamtype=':demote2:' limit 10;");

while (my $row = $db->fetch_array())
{
	print $row, "\n";
}

Common:

my $query = $db->query("select url from edb.white_black_grey where spamtype=':demote2:' limit 10;");

while (my $row = $db->fetch_array($query))
{
	print $row, "\n";
}

=item B<Disconnect>


$db->close();

=back

=head1 DESCRIPTION

B<This lib> will query some information from some different type databases, you can use it very expediently. so we use database easily.

B<This lib> can use dsn which contains all connection infomation or use all single items, like db_host, db_pass etc.

=head1 AUTHOR 

B<Cnangel> (I<junliang.li@alibaba-inc.com>)

=head1 HISTORY

I<Thu Nov  8 13:53:42 2007> B<v0.001> Build.

I<Wed Jan  2 14:54:48 2008> B<v0.002> Build and add postgresql query.

I<Fri Jan 18 15:31:57 2008> B<v0.002> Update.

I<Tue Jun 16 14:52:22 2009> B<v0.003> Build and add oracle query function.

I<Wed Jun 17 10:01:32 2009> B<v0.004> Could use uri which contains connection infomation for databases. 

=cut

